File: tests/docs/c/test_attribute_bindings.cpp
// SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: LicenseRef-NvidiaProprietary
//
// NVIDIA CORPORATION, its affiliates and licensors retain all intellectual
// property and proprietary rights in and to this material, related
// documentation and any modifications thereto. Any use, reproduction,
// disclosure or distribution of this material and related documentation
// without an express license agreement from NVIDIA CORPORATION or
// its affiliates is strictly prohibited.

// Tests for persistent attribute bindings and attribute mapping through the C API.

#include <gtest/gtest.h>
#include "helpers.h"

#include <ovrtx/ovrtx_attributes.h>

#include <cstring>
#include <string>

namespace {

void read_xform_translation_x(ovrtx_renderer_t* renderer, char const* prim_path, double* out_x) {
    ovx_string_t prim = ovx_str(prim_path);
    DLDataType mat_type = {kDLFloat, 64, 16};
    ovrtx_binding_desc_or_handle_t binding =
        ovrtx_make_binding_desc(&prim, 1, ovx_str("omni:xform"), OVRTX_SEMANTIC_NONE, mat_type);

    ovrtx_read_handle_t read_handle = 0;
    ovrtx_enqueue_result_t eq = ovrtx_read_attribute(renderer, &binding, nullptr, &read_handle);
    ASSERT_API_SUCCESS(eq.status);
    docs_wait_no_errors(renderer, eq.op_index);

    ovrtx_read_output_t output{};
    ASSERT_API_SUCCESS(
        ovrtx_fetch_read_result(renderer, read_handle, ovrtx_timeout_infinite, &output).status);
    double const* matrix = static_cast<double const*>(output.buffers[0].dl.data);
    *out_x = matrix[12];

    ovrtx_cuda_sync_t no_sync{};
    ASSERT_API_SUCCESS(ovrtx_release_read_result(renderer, output.map_handle, no_sync).status);
}

void make_xform(double x, double out[16]) {
    std::memset(out, 0, sizeof(double) * 16);
    out[0] = 1.0;
    out[5] = 1.0;
    out[10] = 1.0;
    out[15] = 1.0;
    out[12] = x;
}

} // namespace

class AttributeBindingsTest : public ::testing::Test {
protected:
    static void SetUpTestSuite() {
        TestConfig tc("AttributeBindingsTest");
        ovrtx_result_t result = ovrtx_create_renderer(&tc.config, &renderer_);
        ASSERT_API_SUCCESS(result.status);
    }

    static void TearDownTestSuite() {
        if (renderer_) {
            ovrtx_destroy_renderer(renderer_);
            renderer_ = nullptr;
        }
    }

    static void load_base() {
        ovrtx_enqueue_result_t eq = ovrtx_reset_stage(renderer_);
        ASSERT_API_SUCCESS(eq.status);
        docs_wait_no_errors(renderer_, eq.op_index);

        std::string scene = get_docs_test_data_dir() + "/ovrtx-test-base.usda";
        eq = ovrtx_open_usd_from_file(renderer_, {scene.c_str(), scene.size()});
        ASSERT_API_SUCCESS(eq.status);
        docs_wait_no_errors(renderer_, eq.op_index);

        eq = ovrtx_reset(renderer_, 0.0);
        ASSERT_API_SUCCESS(eq.status);
        docs_wait_no_errors(renderer_, eq.op_index);
    }

    static ovrtx_renderer_t* renderer_;
};

ovrtx_renderer_t* AttributeBindingsTest::renderer_ = nullptr;

TEST_F(AttributeBindingsTest, CreateWriteDestroyBinding) {
    load_base();

    ovx_string_t prim = ovx_str("/World/Plane");
    DLDataType mat_type = {kDLFloat, 64, 16};
    ovrtx_binding_desc_or_handle_t binding_desc =
        ovrtx_make_binding_desc(&prim, 1, ovx_str("omni:xform"), OVRTX_SEMANTIC_XFORM_MAT4x4, mat_type);
    binding_desc.binding_desc.flags = OVRTX_BINDING_FLAG_OPTIMIZE;

    // [snippet:doc-create-attribute-binding-c]
    ovrtx_attribute_binding_handle_t binding_handle = 0;
    ovrtx_enqueue_result_t eq =
        ovrtx_create_attribute_binding(renderer_, &binding_desc.binding_desc, &binding_handle);
    ASSERT_API_SUCCESS(eq.status);
    docs_wait_no_errors(renderer_, eq.op_index);
    // [/snippet:doc-create-attribute-binding-c]

    // [snippet:doc-write-bound-attribute-c]
    ovrtx_binding_desc_or_handle_t binding_ref{};
    binding_ref.binding_handle = binding_handle;

    double matrix[16];
    make_xform(14.0, matrix);
    size_t count = 1;
    DLTensor tensor = ovrtx_make_write_cpu_tensor(matrix, &count, mat_type);
    ovrtx_input_buffer_t buffer{};
    buffer.tensors = &tensor;
    buffer.tensor_count = 1;

    eq = ovrtx_write_attribute(renderer_, &binding_ref, &buffer, OVRTX_DATA_ACCESS_SYNC);
    ASSERT_API_SUCCESS(eq.status);
    docs_wait_no_errors(renderer_, eq.op_index);
    // [/snippet:doc-write-bound-attribute-c]

    double x = 0.0;
    read_xform_translation_x(renderer_, "/World/Plane", &x);
    EXPECT_DOUBLE_EQ(x, 14.0);

    // [snippet:doc-destroy-attribute-binding-c]
    eq = ovrtx_destroy_attribute_binding(renderer_, binding_handle);
    ASSERT_API_SUCCESS(eq.status);
    docs_wait_no_errors(renderer_, eq.op_index);
    // [/snippet:doc-destroy-attribute-binding-c]
}

TEST_F(AttributeBindingsTest, MapAndUnmapAttribute) {
    load_base();

    ovx_string_t prim = ovx_str("/World/Plane");
    DLDataType mat_type = {kDLFloat, 64, 16};
    ovrtx_binding_desc_or_handle_t binding =
        ovrtx_make_binding_desc(&prim, 1, ovx_str("omni:xform"), OVRTX_SEMANTIC_XFORM_MAT4x4, mat_type);

    // [snippet:doc-map-attribute-cpu-c]
    ovrtx_mapping_desc_t mapping_desc{};
    mapping_desc.device_type = kDLCPU;
    mapping_desc.device_id = 0;

    ovrtx_attribute_mapping_t mapping{};
    ovrtx_result_t result = ovrtx_map_attribute(renderer_, &binding, mapping_desc, &mapping);
    ASSERT_API_SUCCESS(result.status);

    double* matrix = static_cast<double*>(mapping.dl.data);
    matrix[0] = 1.0;
    matrix[5] = 1.0;
    matrix[10] = 1.0;
    matrix[15] = 1.0;
    matrix[12] = 15.0;

    ovrtx_cuda_sync_t no_sync{};
    ovrtx_enqueue_result_t eq = ovrtx_unmap_attribute(renderer_, mapping.map_handle, no_sync);
    ASSERT_API_SUCCESS(eq.status);
    docs_wait_no_errors(renderer_, eq.op_index);
    // [/snippet:doc-map-attribute-cpu-c]

    double x = 0.0;
    read_xform_translation_x(renderer_, "/World/Plane", &x);
    EXPECT_DOUBLE_EQ(x, 15.0);
}

File: tests/docs/python/test_attribute_bindings.py
# Copyright (c) 2025-2026, NVIDIA CORPORATION. All rights reserved.
#
# NVIDIA CORPORATION and its licensors retain all intellectual property
# and proprietary rights in and to this software, related documentation
# and any modifications thereto.  Any use, reproduction, disclosure or
# distribution of this software and related documentation without an express
# license agreement from NVIDIA CORPORATION is strictly prohibited.

"""Tests for persistent attribute bindings, mapping, and CUDA write paths."""

from pathlib import Path

import numpy as np
import ovrtx
import warp as wp

TEST_BASE_PATH = str((Path(__file__).parent / "../data/ovrtx-test-base.usda").resolve())


@wp.kernel
def _set_xform_translation_x(transforms: wp.array(dtype=wp.mat44d), x: wp.float64):
    transforms[wp.tid()] = wp.mat44d(
        wp.float64(1.0),
        wp.float64(0.0),
        wp.float64(0.0),
        wp.float64(0.0),
        wp.float64(0.0),
        wp.float64(1.0),
        wp.float64(0.0),
        wp.float64(0.0),
        wp.float64(0.0),
        wp.float64(0.0),
        wp.float64(1.0),
        wp.float64(0.0),
        x,
        wp.float64(0.0),
        wp.float64(0.0),
        wp.float64(1.0),
    )


def _load_base(renderer):
    renderer.open_usd(TEST_BASE_PATH)
    renderer.reset()


def _read_xform(renderer, prim="/World/Plane"):
    tensor = renderer.read_attribute("omni:xform", [prim])
    return np.from_dlpack(tensor).reshape(1, 4, 4)


def _make_xform(x):
    matrix = np.eye(4, dtype=np.float64).reshape(1, 4, 4)
    matrix[0, 3, 0] = x
    return matrix


def test_bind_attribute_write_unbind(renderer):
    """Create a scalar binding, write through it, and read back the result."""
    _load_base(renderer)
    matrix = _make_xform(12.0)

    # [snippet:doc-bind-attribute-write]
    binding = renderer.bind_attribute(
        prim_paths=["/World/Plane"],
        attribute_name="omni:xform",
        dtype="float64",
        shape=(4, 4),
        prim_mode=ovrtx.PrimMode.MUST_EXIST,
        flags=ovrtx.BindingFlag.OPTIMIZE,
    )
    binding.write(matrix)
    binding.unbind()
    # [/snippet:doc-bind-attribute-write]

    np.testing.assert_allclose(_read_xform(renderer), matrix)


def test_bind_attribute_async_and_write_async(renderer):
    """Create and write a binding asynchronously."""
    _load_base(renderer)
    matrix = _make_xform(13.0)

    # [snippet:doc-bind-attribute-async]
    op = renderer.bind_attribute_async(
        ["/World/Plane"], "omni:xform", dtype="float64", shape=(4, 4)
    )
    binding = op.wait()
    assert binding is not None
    # [/snippet:doc-bind-attribute-async]

    # [snippet:doc-binding-write-async]
    write_op = binding.write_async(matrix)
    assert write_op.wait() is True
    # [/snippet:doc-binding-write-async]

    np.testing.assert_allclose(_read_xform(renderer), matrix)
    binding.unbind()


def test_bind_array_attribute(renderer):
    """Bind a variable-length array attribute, write through it, and read it back."""
    _load_base(renderer)
    points = np.array(
        [
            [-1.0, 0.0, -1.0],
            [1.0, 0.0, -1.0],
            [-1.0, 0.0, 1.0],
            [1.0, 0.0, 1.0],
        ],
        dtype=np.float32,
    )

    # [snippet:doc-bind-array-attribute]
    binding = renderer.bind_array_attribute(
        ["/World/Plane"], "points", dtype="float32", shape=(3,)
    )
    binding.write([points])
    # [/snippet:doc-bind-array-attribute]

    values = np.from_dlpack(renderer.read_array_attribute("points", ["/World/Plane"])["/World/Plane"])
    np.testing.assert_allclose(values, points)
    binding.unbind()


def test_map_attribute_cpu(renderer):
    """Map an attribute by name on CPU and verify the mutation."""
    _load_base(renderer)

    # [snippet:doc-map-attribute-cpu]
    with renderer.map_attribute(
        ["/World/Plane"], "omni:xform", dtype="float64", shape=(4, 4)
    ) as mapping:
        matrices = np.from_dlpack(mapping.tensor).reshape(1, 4, 4)
        matrices[0, 3, 0] = 10.0
    # [/snippet:doc-map-attribute-cpu]

    assert _read_xform(renderer)[0, 3, 0] == 10.0


def test_map_bound_attribute_cpu(renderer):
    """Map through an AttributeBinding rather than by name."""
    _load_base(renderer)

    # [snippet:doc-map-bound-attribute]
    binding = renderer.bind_attribute(["/World/Plane"], "omni:xform", dtype="float64", shape=(4, 4))
    with binding.map(device=ovrtx.Device.CPU) as mapping:
        matrices = np.from_dlpack(mapping.tensor).reshape(1, 4, 4)
        matrices[0, 3, 0] = 8.0
    # [/snippet:doc-map-bound-attribute]

    assert _read_xform(renderer)[0, 3, 0] == 8.0
    binding.unbind()


def test_unmap_attribute_async(renderer):
    """Unmap explicitly through the async API."""
    _load_base(renderer)

    # [snippet:doc-unmap-attribute-async]
    mapping = renderer.map_attribute(["/World/Plane"], "omni:xform", dtype="float64", shape=(4, 4))
    matrices = np.from_dlpack(mapping.tensor).reshape(1, 4, 4)
    matrices[0, 3, 0] = 9.0
    op = mapping.unmap_async()
    assert op.wait() is True
    # [/snippet:doc-unmap-attribute-async]

    assert _read_xform(renderer)[0, 3, 0] == 9.0


def test_map_attribute_cuda(renderer):
    """Map an attribute on CUDA, edit it with Warp, and read back on CPU."""
    _load_base(renderer)

    # [snippet:doc-map-attribute-cuda]
    mapping = renderer.map_attribute(
        ["/World/Plane"],
        "omni:xform",
        dtype="float64",
        shape=(4, 4),
        device=ovrtx.Device.CUDA,
    )
    tensor = wp.from_dlpack(mapping.tensor, dtype=wp.mat44d)
    stream = wp.Stream(device=tensor.device)
    wp.launch(_set_xform_translation_x, dim=1, inputs=[tensor, wp.float64(6.0)], stream=stream)
    mapping.unmap(stream=stream.cuda_stream)
    # [/snippet:doc-map-attribute-cuda]

    assert _read_xform(renderer)[0, 3, 0] == 6.0


def test_write_attribute_async_data_access_cuda(renderer):
    """Write a CUDA tensor with asynchronous data access and stream sync."""
    _load_base(renderer)
    cuda_tensor = wp.empty(1, dtype=wp.mat44d, device="cuda:0")
    stream = wp.Stream(device=cuda_tensor.device)
    wp.launch(_set_xform_translation_x, dim=1, inputs=[cuda_tensor, wp.float64(7.0)], stream=stream)

    # [snippet:doc-write-attribute-async-data-access]
    op = renderer.write_attribute_async(
        ["/World/Plane"],
        "omni:xform",
        cuda_tensor,
        data_access=ovrtx.DataAccess.ASYNC,
        cuda_stream=stream.cuda_stream,
    )
    assert op.wait() is True
    # [/snippet:doc-write-attribute-async-data-access]

    assert _read_xform(renderer)[0, 3, 0] == 7.0


def test_write_token_array_attribute(renderer):
    """Write string data as a token array rather than a path relationship."""
    _load_base(renderer)

    # [snippet:doc-write-token-array]
    renderer.write_array_attribute(
        ["/World/Plane"],
        "omni:docTokens",
        [["sensor", "validated"]],
        is_token=True,
        prim_mode=ovrtx.PrimMode.CREATE_NEW,
    )
    # [/snippet:doc-write-token-array]

    prims = renderer.query_prims(
        require_all=[(ovrtx.FilterKind.HAS_ATTRIBUTE, "omni:docTokens")],
        attribute_filter_mode=ovrtx.AttributeFilterMode.SPECIFIC,
        attribute_names=["omni:docTokens"],
    )
    assert "/World/Plane" in prims
    token_info = prims["/World/Plane"]["omni:docTokens"]
    assert token_info.is_array is True
    assert token_info.semantic == ovrtx.Semantic.TOKEN_ID
